src/desktop_file.rs

```diff
@@ -19,55 +37,156 @@ fn find_application_dirs() -> Vec<PathBuf> {
+
+    dekstop_files
+}
+
 /// Attempts to locate a `.desktop` file for an app id
-/// (or app class).
-///
-/// A simple case-insensitive check is performed on filename == `app_id`.
 pub fn find_desktop_file(app_id: &str) -> Option<PathBuf> {
-    let dirs = find_application_dirs();
+    // this is necessary to invalidate the cache
+    let files = find_desktop_files();
 
-    for dir in dirs {
-        let mut walker = WalkDir::new(dir).max_depth(5).into_iter();
+    if let Some(path) = find_desktop_file_by_filename(app_id, &files) {
+        return Some(path);
+    }
 
-        let entry = walker.find(|entry| {
-            entry.as_ref().map_or(false, |entry| {
-                let file_name = entry.file_name().to_string_lossy().to_lowercase();
-                let test_name = format!("{}.desktop", app_id.to_lowercase());
-                file_name == test_name
-            })
-        });
+    let mut desktop_files_cache = lock!(DESKTOP_FILES);
 
-        if let Some(Ok(entry)) = entry {
-            let path = entry.path().to_owned();
+    // invalidate the cache of `desktop_files`
+    if desktop_files_cache.len() != files.len() {
+        desktop_files_cache.clear();
+        get_desktop_files_map(files).clone_into(&mut desktop_files_cache);
+    }
+
+    find_desktop_file_by_filedata(app_id, desktop_files_cache)
+}
+
+/// Finds the correct desktop file using a simple condition check
+fn find_desktop_file_by_filename(app_id: &str, files: &Vec<PathBuf>) -> Option<PathBuf> {
+    let app_id = app_id.to_lowercase();
+
+    for file in files {
+        let file_name: String = file
+            .file_name()
+            .expect("file name doesn't end with ...")
+            .to_string_lossy()
+            .to_lowercase();
+
+        let has_match = file_name.contains(&app_id)
+            || app_id
+                .split(&['-', ' ', ':', '@', '.', '_'][..])
+                .any(|part| file_name.contains(part)); // this will attempt to find flatpak apps that are like this
+                                                       // `com.company.app` or `com.app.something`
+
+        if has_match {
+            let path = file.to_owned();
             return Some(path);
         }
     }
 
     None
 }
 
-/// Parses a desktop file into a flat hashmap of keys/values.
-fn parse_desktop_file(path: PathBuf) -> io::Result<HashMap<String, String>> {
-    let file = File::open(path)?;
-    let lines = io::BufReader::new(file).lines();
+/// Finds the correct desktop file using the keys in `DESKTOP_FILES_LOOK_OUT_KEYS`
+fn find_desktop_file_by_filedata(
+    app_id: &str,
+    desktop_files_map: MutexGuard<'_, HashMap<PathBuf, DesktopFile>>,
+) -> Option<PathBuf> {
+    let app_id = &app_id.to_lowercase();
 
-    let mut map = HashMap::new();
+    let has_match = desktop_files_map.iter().find(|(_, desktop_file)| {
+        desktop_file
+            .values()
+            .flatten()
+            .any(|value| value.to_lowercase().contains(app_id))
+    });
```

Comment:
Here you could map to get the path inline:

```suggestion
    desktop_files_map.iter().find(|(_, desktop_file)| {
        desktop_file
            .values()
            .flatten()
            .any(|value| value.to_lowercase().contains(app_id))
    })
    .map(|(path, _)| path.to_owned())
```

Then remove the `match` underneath.